nodes/node_balancer/cmd/middleware.go

```diff
@@ -31,12 +77,92 @@ func panicMiddleware(next http.Handler) http.Handler {
 // Log access requests in proper format
 func logMiddleware(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, "Unable to read body", http.StatusBadRequest)
+			return
+		}
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		if len(body) > 0 {
+			defer r.Body.Close()
+		}
+
 		next.ServeHTTP(w, r)
+
 		ip, _, err := net.SplitHostPort(r.RemoteAddr)
 		if err != nil {
-			log.Printf("Unable to parse client IP: %s\n", r.RemoteAddr)
+			http.Error(w, fmt.Sprintf("Unable to parse client IP: %s", r.RemoteAddr), http.StatusBadRequest)
+			return
+		}
+		logStr := fmt.Sprintf("%s %s %s", ip, r.Method, r.URL.Path)
+
+		// Parse body and log method if jsonrpc path
+		pathSlice := strings.Split(r.URL.Path, "/")
+		if r.Method == "POST" && pathSlice[len(pathSlice)-1] == "jsonrpc" {
+			var jsonrpcRequest JSONRPCRequest
+			err = json.Unmarshal(body, &jsonrpcRequest)
+			if err != nil {
+				log.Printf("Unable to parse body %v", err)
+			}
+			logStr += fmt.Sprintf(" %s", jsonrpcRequest.Method)
+		}
+
+		if stateCLI.enableDebugFlag {
+			if r.URL.RawQuery != "" {
+				logStr += fmt.Sprintf(" %s", r.URL.RawQuery)
+			}
+			accessID := extractAccessID(r)
+			if accessID != "" {
+				dataSource := extractDataSource(r)
+				logStr += fmt.Sprintf(" %s %s", dataSource, accessID)
+			}
+		}
+		log.Printf("%s\n", logStr)
+	})
+}
+
+// Check access id was provided correctly and save user access configuration to request context
+func accessMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var currentUserAccess UserAccess
+
+		accessID := extractAccessID(r)
+		dataSource := extractDataSource(r)
+
+		if accessID == "" {
+			http.Error(w, "No authorization header passed with request", http.StatusForbidden)
+			return
+		}
+
+		// If access id does not belong to controller, then find it in Bugout resources
+		if accessID == configs.NB_CONTROLLER_ACCESS_ID {
+			currentUserAccess = controllerUserAccess
+			currentUserAccess.dataSource = dataSource
 		} else {
-			log.Printf("%s %s %s\n", ip, r.Method, r.URL.Path)
+			resources, err := bugoutClient.GetResources(configs.NB_CONTROLLER_TOKEN, "", accessID)
```

Comment:
Can you cache this so that we aren't making request to Brood with every access request? Otherwise, it could slow down our crawlers a lot.

Reply:
At least, we should profile/benchmark before we deploy to production as it could affect performance/stability of our crawlers.

Benchmarking can be done with @Yhtiyar 's ERC721 + ERC20 crawler.

Reply:
Implemented fast solution with check crawlers env variable, in futer will be implemented with cache. It requires rework of `Client` and `ClientPool`.